test/unit/runtime/src/KernelThrow.cpp

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2022  René Widera, Mehmet Yusufoglu, Andrea Bocci
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+#include <alpaka/core/RuntimeMacros.hpp>
+#include <alpaka/test/acc/TestAccs.hpp>
+#include <alpaka/test/queue/Queue.hpp>
+
+#include <catch2/catch_template_test_macros.hpp>
+#include <catch2/catch_test_macros.hpp>
+
+class KernelWithThrow
+{
+public:
+    template<typename TAcc>
+    ALPAKA_FN_ACC auto operator()(TAcc const& acc) const -> void
+    {
+        using Idx = alpaka::Idx<TAcc>;
+        using Dim = alpaka::Dim<TAcc>;
+        using Vec = alpaka::Vec<Dim, Idx>;
+        Vec const globalThreadIdx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        if(globalThreadIdx == Vec::zeros())
+        {
+            // Throw abort or std::runtime_error depending on acc type
+            ALPAKA_DEVICE_THROW("Exception thrown by the kernel.");
+        }
+        alpaka::syncBlockThreads(acc);
+    }
+};
+
+template<typename T, typename Acc>
+void checkThrow(std::string const& expectedErrStr)
+{
+    if constexpr(alpaka::accMatchesTags<Acc, T>)
+    {
+        using Idx = alpaka::Idx<Acc>;
+        using Dim = alpaka::Dim<Acc>;
+        using Vec = alpaka::Vec<Dim, Idx>;
+        using Queue = alpaka::Queue<Acc, alpaka::Blocking>;
+
+        auto const platformAcc = alpaka::Platform<Acc>{};
+        auto const devAcc = alpaka::getDevByIdx(platformAcc, 0);
+
+        Queue queue(devAcc);
+        auto const workDiv = alpaka::WorkDivMembers<Dim, Idx>{Vec{8}, Vec{1}, Vec{1}};
+
+        try
+        {
+            alpaka::exec<Acc>(queue, workDiv, KernelWithThrow{});
+            // Cuda can catch exceptions which were thrown at kernel during the wait(); therefore wait is added.
+            alpaka::wait(queue);
+        }
+        catch(std::runtime_error& e)
+        {
+            std::string const errorStr{e.what()};
+            printf("The error str catched: %s \n", errorStr.c_str());
+            printf("The expected str in error str: %s \n", expectedErrStr.c_str());
+
+            auto const found = errorStr.find(expectedErrStr);
+            CHECK(found != std::string::npos);
+        }
+        catch(std::exception& e)
+        {
+            FAIL(std::string("Wrong exception type thrown in kernel:") + e.what());
+        }
+    }
+}
+
+using TestAccs = alpaka::test::EnabledAccs<alpaka::DimInt<1u>, std::uint32_t>;
+
+TEMPLATE_LIST_TEST_CASE("ThrowForCpuThreadAndSerial", "[runtime]", TestAccs)
+{
+    using Acc = TestType;
+    // Test runtime-error exceptions. Aborts can not be catched.
+    checkThrow<alpaka::TagCpuThreads, Acc>("Exception thrown by the kernel");
+    checkThrow<alpaka::TagCpuSerial, Acc>("Exception thrown by the kernel");
+    checkThrow<alpaka::TagGpuCudaRt, Acc>("cudaErrorLaunchFailure");
```

Comment:
can you add the tests for the other backends ?

Reply:
@fwyzard  can you test hip/rocm abort() if not difficult for you, or if you don't mind. I tested but can not trigger abort with abort() function for only a specific setup. ( For `OpenMp` I explained the issue at PR explanation, the `sycl` assert(false) can not be catched. Only `Hip` can be added but I assumed it is not catchable somehow, may be due to Rene's explanations. )

Reply:
`__assert_fail` did not work for hip/rocm.

Reply:
`__assert_fail()` is identical to `abort()`

Reply:
Thanks a lot, good info.  I tried  `asm volatile("s_trap 0x1;"); return; ` but that did not stop the kernel. There is no `raise` function to `raise(SIGABRT).` Null pointer access to create _segmentation fault_ stops the kernel, then host program stalls, but does not exit.

Reply:
` __builtin_trap()` is recommended by  @psychocoderHPC  . I will try.

Reply:
Andrea already tried and said it behaved like `abort()`. Looking at the implementation they are the same:
```cpp
// abort
extern "C" __device__ inline __attribute__((weak))
void abort() {
  __builtin_trap();
}
```

Reply:
Yes, I used assert(false) eventually.